isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java

```diff
@@ -0,0 +1,159 @@
+  }
+
+  @Override
+  public RexNode visit(Expression.TimeLiteral expr) throws RuntimeException {
+    // Expression.TimeLiteral is Micros, while RexLiteral assumes Milliseconds for time type
+    int milliS = (int) (TimeUnit.MICROSECONDS.toMillis(expr.value()));
+    int nanoS =
+        (int) (TimeUnit.MICROSECONDS.toNanos(expr.value()) - TimeUnit.MILLISECONDS.toNanos(milliS));
+
+    TimeString timeString = TimeString.fromMillisOfDay(milliS).withNanos(nanoS);
+    return rexBuilder.makeLiteral(timeString, TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.DateLiteral expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.TimestampLiteral expr) throws RuntimeException {
+    // Expression.TimestampLiteral is Micros, while RexLiteral assumes Milliseconds for timestamp
+    // type
+    long milliS = TimeUnit.MICROSECONDS.toMillis(expr.value());
+    int nanoS =
+        (int) (TimeUnit.MICROSECONDS.toNanos(expr.value()) - TimeUnit.MILLISECONDS.toNanos(milliS));
+
+    TimestampString tsString = TimestampString.fromMillisSinceEpoch(milliS).withNanos(nanoS);
+    return rexBuilder.makeLiteral(tsString, TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.TimestampTZLiteral expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.ScalarFunctionInvocation expr) throws RuntimeException {
+    var args = expr.arguments().stream().map(a -> a.accept(this)).toList();
+    Optional<SqlOperator> operator = scalarFunctionConverter.getSqlOperatorFromSubstraitFunc(expr);
+    if (operator.isPresent()) {
+      return rexBuilder.makeCall(operator.get(), args);
+    }
+
+    return visitFallback(expr);
+  }
+
+  @Override
+  public RexNode visit(Expression.Cast expr) throws RuntimeException {
+    return rexBuilder.makeAbstractCast(
+        TypeConverter.convert(typeFactory, expr.getType()), expr.input().accept(this));
+  }
+
+  @Override
+  public RexNode visit(FieldReference expr) throws RuntimeException {
+    if (expr.isSimpleRootReference()) {
+      var segment = expr.segments().get(0);
+
+      RexInputRef rexInputRef =
+          switch (segment) {
+            case FieldReference.StructField f -> {
+              yield new RexInputRef(f.offset(), TypeConverter.convert(typeFactory, expr.getType()));
+            }
+            default -> throw new IllegalArgumentException("Unhandled type: " + segment);
+          };
+
+      return rexInputRef;
+    }
+
+    return visitFallback(expr);
+  }
+
+  @Override
+  public RexNode visitFallback(Expression expr) {
+    throw new UnsupportedOperationException(
+        String.format(
```

Comment:
You might want to include the actual expression, not just the class. For example, a fieldReference that includes an array index wouldn't be obvious from this message (I don't think).

Reply:
I added actual expression in the Exception message.